packages/server/src/api/controllers/row/staticFormula.js

```diff
@@ -0,0 +1,156 @@
+const CouchDB = require("../../../db")
+const { getRowParams } = require("../../../db/utils")
+const {
+  outputProcessing,
+  processAutoColumn,
+  processFormulas,
+} = require("../../../utilities/rowProcessor")
+const { FieldTypes, FormulaTypes } = require("../../../constants")
+const { isEqual } = require("lodash")
+const { cloneDeep } = require("lodash/fp")
+
+/**
+ * This function runs through a list of enriched rows, looks at the rows which
+ * are related and then checks if they need the state of their formulas
+ * updated.
+ * NOTE: this will only for affect static formulas.
+ */
+exports.updateRelatedFormula = async (appId, table, enrichedRows) => {
+  const db = new CouchDB(appId)
+  // no formula to update, we're done
+  if (!table.relatedFormula) {
+    return
+  }
+  let promises = []
+  for (let enrichedRow of Array.isArray(enrichedRows)
+    ? enrichedRows
+    : [enrichedRows]) {
+    // the related rows by tableId
+    let relatedRows = {}
+    for (let [key, field] of Object.entries(enrichedRow)) {
+      const columnDefinition = table.schema[key]
+      if (columnDefinition && columnDefinition.type === FieldTypes.LINK) {
+        const relatedTableId = columnDefinition.tableId
+        if (!relatedRows[relatedTableId]) {
+          relatedRows[relatedTableId] = []
+        }
+        relatedRows[relatedTableId] = relatedRows[relatedTableId].concat(field)
+      }
+    }
+    for (let tableId of table.relatedFormula) {
+      try {
+        // no rows to update, skip
+        if (!relatedRows[tableId] || relatedRows[tableId].length === 0) {
+          continue
+        }
+        const relatedTable = await db.get(tableId)
+        for (let column of Object.values(relatedTable.schema)) {
+          // needs updated in related rows
+          if (
+            column.type === FieldTypes.FORMULA &&
+            column.formulaType === FormulaTypes.STATIC
+          ) {
+            // re-enrich rows for all the related, don't update the related formula for them
+            promises = promises.concat(
+              relatedRows[tableId].map(related =>
+                exports.finaliseRow(appId, relatedTable, related, {
+                  updateFormula: false,
+                })
+              )
+            )
+          }
+        }
+      } catch (err) {
+        // no error scenario, table doesn't seem to exist anymore, ignore
```

Comment:
Is there anything more specific we can do here to ensure this is why the error has been thrown? There's a fair bit of logic in `finaliseRow` that could go wrong and give us a false ignore here.

If there is an error here that we shouldn't ignore I wonder what's the best approach, perhaps to continue with the updates as much as possible but flag that the operation was 'completed with some errors' and flag that to the user?

Reply:
Good catch about the `finaliseRow` errors, I think in this error scenario it will need to just error as we have no real way to alert the user to a partial completion. This is an issue that occurs in a few places, like bulk deletions since we have no way to perform database ACID transactions.

This is something that we could address with the Budibase bulk/long running/transaction system, with bulk updates have a consistent approach and API response for when errors occur.

Reply:
Alerts aside, is there anything we can do in the mean time to check this error was caused by the table not existing anymore, do we get a 404 response anywhere?

Reply:
For alerts in the mean time, it would be possible to allow the iteration to complete and mark non expected errors in an array. Then do a final throw if there are present, so that we can return those errors to the UI

Reply:
For now I'll just move the finalise rows out of the try catch, that way it'll only catch the error of the table not existing - its an edge case scenario that shouldn't occur but if it does then we can just move on, the relationship isn't present anymore.

As for alerts, I think its probably not the best idea to do something about this specific case, as its a problem in multiple places, its probably best to try and resolve the issue of bulk operations having error scenarios in one go, this applies to lots of things, like adding/updating relationships, bulk deleting rows, adding/removing columns from a table etc.

Reply:
That should cover it then as it'll error out on any other scenario, we shouldn't need the alert then. 

All for a global strategy for error handling in bulk ops, was just mentioning above so we can get some form of protection in the mean time. Not even sure it needs to be coupled to the worker piece, we could even add some wrapper behaviour like:

```js
/**
* Perform an operation that impacts many rows. 
* Because CouchDB does not support ACID transactions there
* needs to be some allowance for failures, in which case we will notify the user 
* that the operation was not completely successful 
*/
bulkOperation = async (ctx, operation) => {
   try {
      await operation() 
   } catch (err) {
      if (err.partialError) {
         // don't completely fail
      }
      if (err.errors) {
         // add to body
      }
      if (err.status) {
        // set the desired status
      }
      if (err.message) {
        // add to body
      }
   }
```

or something similar. I can move this out to a separate task for now and reference this PR as an example of somewhere we could benefit.

Reply:
Added: https://github.com/Budibase/budibase/issues/4270